include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,328 @@
+{
+	public:
+		virtual ~CNormalMapToDerivativeFilter() {}
+
+		class CStateBase : public CMatchedSizeInOutImageFilterCommon::state_type, public CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase
+		{ 
+			public:
+				CStateBase() = default;
+				virtual ~CStateBase() = default;
+
+			private:
+
+				void setLayerScaleValuesOffset()
+				{
+					scaleValuesPointer = reinterpret_cast<float*>(scratchMemory) + (extent.width * extent.height * extent.depth * forcedScratchChannelAmount);
+				}
+
+				void resetLayerScaleValues()
+				{
+					memset(const_cast<float*>(scaleValuesPointer), 0, layerCount * forcedScratchChannelAmount * decodeTypeByteSize);
+				}
+
+				friend class CNormalMapToDerivativeFilter;
+		};
+		using state_type = CStateBase; //!< full combined state
+
+		static inline bool validate(state_type* state)
+		{
+			if (!CMatchedSizeInOutImageFilterCommon::validate(state))
+				return false;
+
+			if (!CNormalMapToDerivativeFilterBase<Swizzle>::validate(state))
+				return false;
+			
+			const ICPUImage::SCreationParams& inParams = state->inImage->getCreationParameters();
+			const ICPUImage::SCreationParams& outParams = state->outImage->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+
+			if (state->scratchMemoryByteSize < state_type::getRequiredScratchByteSize(state->layerCount, state->extent))
+				return false;
+
+			if (asset::getFormatChannelCount(inFormat) < 3 && asset::getFormatChannelCount(outFormat) != 2)
+				return false;
+
+			if (asset::isIntegerFormat(inFormat) || asset::isIntegerFormat(outFormat))
+				return false;
+
+			// TODO: remove this later when we can actually write/encode to block formats
+			if (asset::isBlockCompressionFormat(outFormat))
+				return false;
+
+			return true;
+		}
+
+		static inline bool execute(state_type* state)
+		{
+			if (!validate(state))
+				return false;
+
+			state->setLayerScaleValuesOffset();
+			state->resetLayerScaleValues();
+
+			const asset::E_FORMAT inFormat = state->inImage->getCreationParameters().format;
+			const asset::E_FORMAT outFormat = state->outImage->getCreationParameters().format;
+			const auto inTexelByteSize = asset::getTexelOrBlockBytesize(inFormat);
+			const auto outTexelByteSize = asset::getTexelOrBlockBytesize(outFormat);
+			const auto currentChannelCount = asset::getFormatChannelCount(inFormat);
+			const auto arrayLayers = state->inImage->getCreationParameters().arrayLayers;
+			static constexpr auto maxChannels = 4u;
+
+			#ifdef _NBL_DEBUG
+			memset(state->scratchMemory, 0, state->scratchMemoryByteSize);
+			#endif // _NBL_DEBUG
+
+			const core::vector3du32_SIMD scratchByteStrides = TexelBlockInfo(asset::E_FORMAT::EF_R32G32_SFLOAT).convert3DTexelStridesTo1DByteStrides(state->extentLayerCount);
+			const auto scratchTexelByteSize = scratchByteStrides[0];
```

Comment:
your scratch should be a tightly packed (strides equal multiples of dimensions) and a plain array of floats, this has nothing to do with texel blocks

Reply:
it is tightly packed since the equation for scratch memory is

```C++
static inline size_t getRequiredScratchByteSize(size_t layerCount, asset::VkExtent3D extent)
				{
					size_t retval = extent.width * extent.height * extent.depth * decodeTypeByteSize * forcedScratchChannelAmount + (layerCount * decodeTypeByteSize * forcedScratchChannelAmount);
					
					return retval;
				}
```
as far as I can remember we have also assumed the `decode type` is float and `forcedScratchChannelAmount` is 2, so I use the `scratchByteStrides` with the appropriate assumptions to make sure offsets are calculated properly using dedicate local byte offset call   

```C++
const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
								float* data = reinterpret_cast<float*>(state->scratchMemory + offset);
```

correct me if I'm wrong please

Reply:
ah I overlooked that the argument is `state->extentLayerCount`